modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/baseline/IgniteNodeAttributesInBaselineTest.java

```diff
@@ -0,0 +1,171 @@
+
+        cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+            .setDefaultDataRegionConfiguration(
+                new DataRegionConfiguration()
+                    .setMaxSize(256L * 1024 * 1024)
+                    .setPersistenceEnabled(true)));
+
+        Map<String, String> userAttrs = new HashMap<>();
+
+        if (addBasicAttrs)
+            userAttrs.put(BASIC_USER_ATTR_NAME, BASIC_USER_ATTR_VAL);
+
+        if (addExtraAttr)
+            userAttrs.put(EXTRA_USER_ATTR_NAME, EXTRA_USER_ATTR_VAL);
+
+        cfg.setAffinityAttributes(userAttrs);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        cleanPersistenceDir();
+
+        IgniteEx ignite = startGrids(GRIDS_NUM);
+
+        ignite.cluster().state(ACTIVE);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        addBasicAttrs = true;
+
+        addExtraAttr = false;
+
+        super.afterTest();
+    }
+
+    /**
+     * Checks that in case of node joining baseline topology and containing new attribute, this attribute will be
+     * propagated to all nodes' baseline topologies.
+     *
+     * @throws Exception If failed to start node.
+     */
+    @Test
+    public void testNewAttribute() throws Exception {
+        stopGrid(GRIDS_NUM - 1);
+
+        addExtraAttr = true;
+
+        IgniteEx lastGrid = startGrid(GRIDS_NUM - 1);
+
+        Object consId2 = lastGrid.localNode().consistentId();
+
+        for (int i = 0; i < GRIDS_NUM; i++) {
+            Collection<BaselineNode> baselineNodes = grid(i).cluster().currentBaselineTopology();
+
+            BaselineNode bsNode = baselineNodes.stream()
+                .filter(node -> node.consistentId().equals(consId2))
+                .findAny()
+                .orElse(null);
+
+            assertNotNull(bsNode);
+
+            String userAttr = bsNode.affinityAttribute(EXTRA_USER_ATTR_NAME);
+
+            assertEquals(EXTRA_USER_ATTR_VAL, userAttr);
+        }
+    }
+
+    /**
+     * Checks that in case of node joining baseline topology and lacking attribute that is present in other nodes'
+     * baseline topologies, this node will not be allowed to join the grid.
+     */
+    @Test
+    public void testExistingAttribute() {
+        stopGrid(GRIDS_NUM - 1);
+
+        addBasicAttrs = false;
+
+        try {
+            startGrid(GRIDS_NUM - 1);
+        }
+        catch (Exception e) {
+            IgniteCheckedException cause = (IgniteCheckedException)e.getCause();
+
+            assertNotNull(cause);
+
+            IgniteSpiException rootCause = (IgniteSpiException)cause.getCause();
+
+            assertNotNull(rootCause);
+
+            String msg = rootCause.getMessage();
+
+            String expMsg = "Some affinity attributes in cluster baseline topology are missing on joining node. " +
```

Comment:
i suppose that this string need to be statically coded into appropriate class, additionally seems we need to reformat _Attr name:_ and _Attr val:_ strings - make it more readable.

Reply:
Fixed readability a bit. I deliberately avoided usual "attrName=attrVal" notation because here main peace of information is attrName and attrVal is  optional